### PR TITLE
Add queries for correct use of MDL opaque fields (C28145)

### PR DESCRIPTION
**Pull request by NateD-MSFT**
Code Analysis has the following Must-Fix warning: [C28145 warning - Windows drivers | Microsoft Docs](https://docs.microsoft.com/mt-mt/windows-hardware/drivers/devtest/28145-opaque-mdl-structure-should-not-be-modified)

> The driver code is changing a member of an MDL structure.
> 
> The MdlFlags field is used as a proxy for all MDL fields. No fields should be modified, with the exception of MDL_MAPPING_CAN_FAIL, which is used for drivers that need to be Microsoft Windows 98 or Windows NT (SP4) compatible, and MDL_PAGES_LOCKED, which is used for drivers that need to be Windows 2000 compatible.

And [the MDL documentation on MSDN states](https://docs.microsoft.com/en-us/windows-hardware/drivers/kernel/using-mdls):

> An MDL structure is semi-opaque. Your driver should directly access only the Next and MdlFlags members of this structure. For a code example that uses these two members, see the following Example section.
> 
> The remaining members of an MDL are opaque. Do not access the opaque members of an MDL directly. Instead, use the following macros, which the operating system provides to perform basic operations on the structure:

As such, this PR adds two queries:

- OpaqueMDLUse, which checks for accesses to an opaque MDL field outside of specified macros.  This is not a candidate for CodeQL must-fix as the impact is low.

- OpaqueMDLWrite, which ports and expands C28145 and looks for writes to opaque MDL fields, which carries a higher chance of corruption.  This is a future candidate for CodeQL must-fix to match our CA behavior.

Additionally, this PR tweaks @t-abtamene's test scripts to clean out the "AnalysisFiles" and "TestDB" intermediate directories used when running unit tests.  It also adds the "AnalysisFiles" and "TestDB" folders to the .gitignore file.

### CODE REVIEW

File: src/drivers/wdm/queries/OpaqueMdlWrite/driver_snippet.c [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+_Use_decl_annotations_
+    NTSTATUS
+    DispatchSetInformation(
+        PDEVICE_OBJECT DeviceObject,
+        PIRP Irp)
+{
+    UNREFERENCED_PARAMETER(DeviceObject);
+    UNREFERENCED_PARAMETER(Irp);
+    return STATUS_SUCCESS;
+}
```
Comment (t-abtamene) [inline src/drivers/wdm/queries/OpaqueMdlWrite/driver_snippet.c:33]: Remove this section as it's unrelated to the specific test.

File: src/drivers/wdm/queries/OpaqueMdlUse/driver_snippet.c [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+_Use_decl_annotations_
+    NTSTATUS
+    DispatchSetInformation(
+        PDEVICE_OBJECT DeviceObject,
+        PIRP Irp)
+{
+    UNREFERENCED_PARAMETER(DeviceObject);
+    UNREFERENCED_PARAMETER(Irp);
+    return STATUS_SUCCESS;
+}
```
Comment (t-abtamene) [inline src/drivers/wdm/queries/OpaqueMdlUse/driver_snippet.c:35]: Remove this section as it's unrelated to this specific test.

**Review by t-abtamene — APPROVED** [review]
You can remove this section and replace it with one Macro definition. The way I did it was by having 
`#define SET_PENDING 0` in this driver_snippet.c file and checking for its value in the template file like this
`    #if SET_PENDING == 1
    DriverObject->MajorFunction[IRP_MJ_WRITE]            = (PDRIVER_DISPATCH)DispatchWrite;
    DriverObject->MajorFunction[IRP_MJ_SET_INFORMATION]  = (PDRIVER_DISPATCH)DispatchSetInformation;
    #endif
`
I think it makes it easier to read.

**Comment by NateD-MSFT** [comment]
> You can remove this section and replace it with one Macro definition. The way I did it was by having `#define SET_PENDING 0` in this driver_snippet.c file and checking for its value in the template file like this ` #if SET_PENDING == 1 DriverObject->MajorFunction[IRP_MJ_WRITE] = (PDRIVER_DISPATCH)DispatchWrite; DriverObject->MajorFunction[IRP_MJ_SET_INFORMATION] = (PDRIVER_DISPATCH)DispatchSetInformation; #endif` I think it makes it easier to read.

@t-abtamene Definitely, but I assumed you were going to check in that change to the template file with your next PR and I didn't want to cause merge conflicts or confusion.  Once you make that change to the template I'll go back and update these files to clean them up.

File: src/drivers/wdm/queries/OpaqueMdlWrite/driver_snippet.c (hunk 1 above)
Comment (NateD-MSFT) [reply src/drivers/wdm/queries/OpaqueMdlWrite/driver_snippet.c:33]: See general comment on the PR replying to your review comment.

File: src/drivers/wdm/queries/OpaqueMdlUse/OpaqueMdlUse.ql [hunk 3]
```diff
@@ -0,0 +1,71 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+/**
+ * @name Direct access of opaque MDL field
+ * @description Direct access of opaque MDL fields should be avoided, as opaque struct layouts may change without warning.
```
Comment (lauragonzalezzz) [inline src/drivers/wdm/queries/OpaqueMdlUse/OpaqueMdlUse.ql:5]: LOVE THIS 😃

File: src/drivers/wdm/queries/OpaqueMdlUse/OpaqueMdlUse.sarif [hunk 4]
```diff
@@ -0,0 +1,627 @@
+{
+  "$schema" : "https://raw.githubusercontent.com/oasis-tcs/sarif-spec/master/Schemata/sarif-schema-2.1.0.json",
```
Comment (lauragonzalezzz) [inline src/drivers/wdm/queries/OpaqueMdlUse/OpaqueMdlUse.sarif:2]: Is this a results file? If so, please name accordingly and describe what this is the results of.

**Review by lauragonzalezzz — APPROVED** [review]

File: src/drivers/wdm/queries/OpaqueMdlUse/OpaqueMdlUse.sarif (hunk 4 above)
Comment (NateD-MSFT) [reply src/drivers/wdm/queries/OpaqueMdlUse/OpaqueMdlUse.sarif:2]: Yep, this is the results for the unit tests in driver_snippet.c.  At the moment Abenezer's script is set up to look for a SARIF file with the same name as the query - we can talk about changing the naming style going forward.

**Pull request merged**